src/opm/output/eclipse/WriteRFT.cpp

```diff
@@ -256,26 +344,621 @@ namespace {
+            this->verifyValid(binId);
+
+            return { this->start(binId), this->start(binId + 1) };
+        }
+
+        bool empty(const int binId) const
+        {
+            this->verifyValid(binId);
+
+            return this->start(binId) == this->start(binId + 1);
+        }
+
+        std::optional<int> last(const int binId) const
+        {
+            this->verifyValid(binId);
+
+            if (this->empty(binId)) {
+                return {};
+            }
+
+            auto end = this->start(binId + 1);
+            return *--end;
+        }
+
+    private:
+        int minId_{std::numeric_limits<int>::max()};
+        int maxId_{std::numeric_limits<int>::min()};
+
+        std::vector<std::vector<int>::size_type> pos_{};
+        std::vector<int> ix_{};
+
+        std::vector<int>::const_iterator start(const int binId) const
+        {
+            return this->ix_.begin() + this->pos_[binId - this->minId_];
+        }
+
+        bool valid(const int binId) const
+        {
+            return (binId >= this->minId_)
+                && (binId <= this->maxId_);
+        }
+
+        void verifyValid(const int binId) const
+        {
+            if (this->valid(binId)) {
+                return;
+            }
+
+            throw std::invalid_argument {
+                fmt::format("Bin ID {} outside valid range {}..{}",
+                            binId, this->minId_, this->maxId_)
+            };
+        }
+    };
+
+    template <typename Cmp>
+    void CSRIndexRelation::build(const std::size_t              size,
+                                 const int                      minId,
+                                 const std::function<int(int)>& binId,
+                                 Cmp&&                          cmp)
+    {
+        if (size == std::size_t{0}) {
+            return;
+        }
+
+        this->ix_.resize(size);
+        std::iota(this->ix_.begin(), this->ix_.end(), 0);
+        std::sort(this->ix_.begin(), this->ix_.end(), std::forward<Cmp>(cmp));
+
+        // Sort must respect binId(i1) <= binId(i2)
+        auto inconsistentId =
+            std::adjacent_find(this->ix_.begin(),
+                               this->ix_.end(),
+                [&binId](const int i1, const int i2)
+            {
+                return binId(i1) > binId(i2);
+            });
+
+        if (inconsistentId != this->ix_.end()) {
+            throw std::invalid_argument {
+                "Comparison operator does not honour bin consistency requirement"
+            };
+        }
+
+        const auto binIdBounds =
+            std::minmax_element(this->ix_.begin(), this->ix_.end(),
+                                [&binId](const int i1, const int i2)
+                                { return binId(i1) < binId(i2); });
+
+        if (binIdBounds.first != this->ix_.end()) {
+            this->minId_ = std::min(binId(*binIdBounds.first), minId);
+        }
+
+        if (binIdBounds.second != this->ix_.end()) {
```

Comment:
Seems to be implied by the previous if. If there is a min there has to be a max

Reply:
> Seems to be implied by the previous if. If there is a min there has to be a max.

Good point.  I've fused the two conditionals.